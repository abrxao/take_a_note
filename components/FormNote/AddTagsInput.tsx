"use client";
import { ChangeEventHandler, MouseEventHandler, ReactNode } from "react";
import { Label } from "../ui/label";
import { Input } from "../ui/input";
import { Button } from "../ui/button";
import { Plus } from "lucide-react";
import TagsArea from "./TagsArea";

interface AddTagsInput {
  tags: string[];
  addButtonAction?: MouseEventHandler<HTMLButtonElement>;
  inputOnChange?: ChangeEventHandler<HTMLInputElement>;
  inputValue?: string | number | readonly string[];
  clickOnTagAction?: MouseEventHandler<HTMLButtonElement>;
}

export default function AddTagsInput({
  tags,
  addButtonAction,
  inputOnChange,
  inputValue,
  clickOnTagAction,
}: AddTagsInput): ReactNode {
  return (
    <div className="grid gap-2">
      <Label htmlFor="addTags">Add tags</Label>

      <div>
        <p className="text-xs text-zinc-600 mt-1 dark:text-zinc-300">
          Your tags
        </p>
        <TagsArea
          tags={tags}
          Icon={<Plus size={14} />}
          clickAction={clickOnTagAction}
        />
      </div>
      <div className="relative">
        <Input
          id="addTags"
          placeholder="New Tag"
          onChange={inputOnChange}
          type="text"
          value={inputValue}
          className="my-0"
        />
        <Button
          type="button"
          className="absolute right-0 top-0"
          onClick={addButtonAction}
        >
          <Plus size={24} />
        </Button>
      </div>
    </div>
  );
}
