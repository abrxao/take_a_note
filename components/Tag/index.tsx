"use client";
import { ButtonHTMLAttributes, ReactNode } from "react";
import { Badge } from "../ui/badge";
import { twMerge } from "tailwind-merge";

interface TagProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  clickAction?: () => void;
}

export default function Tag({
  children,
  className,
  disabled,
  clickAction,
  ...props
}: TagProps): ReactNode {
  return (
    <button
      {...props}
      disabled={disabled || clickAction == undefined}
      className={twMerge("hover:opacity-85", className)}
      onClick={() => clickAction && clickAction()}
    >
      <Badge className="flex gap-1 items-center pointer-events-none">
        {children}
      </Badge>
    </button>
  );
}
