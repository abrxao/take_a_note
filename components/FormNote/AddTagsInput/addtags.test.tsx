import { act, fireEvent, render, screen } from "@testing-library/react";
import AddTagsInput from ".";
import { useState } from "react";
import Tag from "@/components/Tag";
import { X } from "lucide-react";

function AddTagsSimulationComponent() {
  const [tags, setTags] = useState<string[]>(["lorem", "ipsum"]);
  const [selectedTags, setSelectedTags] = useState<string[]>([]);
  const [currentTag, setCurrentTag] = useState<string>("");

  return (
    <>
      <AddTagsInput
        inputOnChange={(e) => setCurrentTag(e.target.value)}
        inputValue={currentTag}
        tags={tags}
        clickOnTagAction={(tag: string) => {
          const aux = tags;
          aux.push(tag);
          setTags(aux);
        }}
        addButtonAction={() => {
          const aux = tags;
          aux.push(currentTag);
          setTags(aux);
        }}
      />
      <ul className={"flex flex-wrap items-start space-x-2"}>
        {selectedTags.map((elem) => {
          return (
            <li key={elem}>
              <Tag
                type="button"
                value={elem}
                aria-label="remove tag"
                clickAction={() => {
                  // @ts-ignore
                  const aux = selectedTags.filter((tag) => tag != elem);
                  setTags(aux);
                }}
              >
                {elem}
                <X size={14} />
              </Tag>
            </li>
          );
        })}
      </ul>
    </>
  );
}

test("simulation input add tag", async () => {
  const { container } = render(<AddTagsSimulationComponent />);
  const newTag = "new tag 2";
  const inputAddTags = screen.getByTestId("input-add-tags");
  const button = screen.getByTestId("input-add-tags-button");
  await act(async () => {
    fireEvent.change(inputAddTags, { target: { value: newTag } });
    fireEvent.click(button);
  });
  expect(container).toMatchSnapshot();
  expect(inputAddTags).toBeInTheDocument();
});
