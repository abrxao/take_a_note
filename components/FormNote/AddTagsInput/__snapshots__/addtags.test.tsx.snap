// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simulation input add tag 1`] = `
<div>
  <div
    class="grid gap-2"
  >
    <label
      class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
      for="addTags"
    >
      Add tags
    </label>
    <div>
      <p
        class="text-xs text-zinc-600 mt-1 dark:text-zinc-300"
      >
        Your tags
      </p>
      <ul
        class="flex flex-wrap items-start space-x-2"
      >
        <li>
          <button
            aria-label="remove tag"
            class="hover:opacity-85"
            type="button"
            value="lorem"
          >
            <div
              class="rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-slate-950 focus:ring-offset-2 dark:border-slate-800 dark:focus:ring-slate-300 border-transparent bg-slate-900 text-slate-50 hover:bg-slate-900/80 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/80 flex gap-1 items-center pointer-events-none"
            >
              lorem
              <svg
                class="lucide lucide-plus "
                fill="none"
                height="14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
            </div>
          </button>
        </li>
        <li>
          <button
            aria-label="remove tag"
            class="hover:opacity-85"
            type="button"
            value="ipsum"
          >
            <div
              class="rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-slate-950 focus:ring-offset-2 dark:border-slate-800 dark:focus:ring-slate-300 border-transparent bg-slate-900 text-slate-50 hover:bg-slate-900/80 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/80 flex gap-1 items-center pointer-events-none"
            >
              ipsum
              <svg
                class="lucide lucide-plus "
                fill="none"
                height="14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
            </div>
          </button>
        </li>
      </ul>
    </div>
    <div
      class="relative"
    >
      <input
        class="flex h-10 w-full rounded-md border border-slate-200 bg-white px-3 py-2 text-sm ring-offset-white file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-slate-500 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-800 dark:bg-slate-950 dark:ring-offset-slate-950 dark:placeholder:text-slate-400 dark:focus-visible:ring-slate-300 my-0"
        data-testid="input-add-tags"
        id="addTags"
        placeholder="New Tag"
        value="new tag 2"
      />
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-slate-900 text-slate-50 hover:bg-slate-900/90 dark:bg-slate-50 dark:text-slate-900 dark:hover:bg-slate-50/90 h-10 px-4 py-2 absolute right-0 top-0"
        data-testid="input-add-tags-button"
        type="button"
      >
        <svg
          class="lucide lucide-plus "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
      </button>
    </div>
  </div>
  <ul
    class="flex flex-wrap items-start space-x-2"
  />
</div>
`;
